From: Alan Shearer <alan@localhost>
Date: Tue, 22 Sep 2026 04:38:00 -0500
Subject: [PATCH] virtio_block: check get_parent_node / get_driver

virtio_block_init_device immediately calls negotiate_features on
info->virtio. If get_parent_node or get_driver fails, that pointer is
NULL and the next call KDLs.

---
 src/add-ons/kernel/drivers/disk/virtual/virtio_block/virtio_block.cpp | 14 +++++++++++---
 1 file changed, 11 insertions(+), 3 deletions(-)

diff --git a/src/add-ons/kernel/drivers/disk/virtual/virtio_block/virtio_block.cpp b/src/add-ons/kernel/drivers/disk/virtual/virtio_block/virtio_block.cpp
index 9eb71254daec..d3c0d0d0d0d0 100644
--- a/src/add-ons/kernel/drivers/disk/virtual/virtio_block/virtio_block.cpp
+++ b/src/add-ons/kernel/drivers/disk/virtual/virtio_block/virtio_block.cpp
@@ -262,11 +262,19 @@ virtio_block_init_device(void* _info, void** _cookie)
 	CALLED();
 	virtio_block_driver_info* info = (virtio_block_driver_info*)_info;
 
 	device_node* parent = sDeviceManager->get_parent_node(info->node);
+	if (parent == NULL)
+		return B_BAD_VALUE;
+
-	sDeviceManager->get_driver(parent, (driver_module_info **)&info->virtio,
+	status_t status = sDeviceManager->get_driver(parent,
+		(driver_module_info **)&info->virtio,
 		(void **)&info->virtio_device);
 	sDeviceManager->put_node(parent);
+	if (status != B_OK)
+		return status;
+	if (info->virtio == NULL)
+		return B_BAD_VALUE;
 
 	info->virtio->negotiate_features(info->virtio_device,
 		VIRTIO_BLK_F_BARRIER | VIRTIO_BLK_F_SIZE_MAX
 			| VIRTIO_BLK_F_SEG_MAX | VIRTIO_BLK_F_GEOMETRY
 			| VIRTIO_BLK_F_RO | VIRTIO_BLK_F_BLK_SIZE
 			| VIRTIO_BLK_F_FLUSH | VIRTIO_BLK_F_TOPOLOGY
 			| VIRTIO_FEATURE_RING_INDIRECT_DESC,
 		&info->features, &get_feature_name);
 
-	status_t status = info->virtio->read_device_config(
+	status = info->virtio->read_device_config(
 		info->virtio_device, 0, &info->config,
 		sizeof(struct virtio_blk_config));
 	if (status != B_OK)
 		return status;
-- 
2.45.2
